README
AACRAID Driver for Linux (take two)

Introduction
-------------------------
The aacraid driver adds support for Adaptec (http://www.adaptec.com)
RAID controllers. This is a major rewrite from the original
Adaptec supplied driver. It has significantly cleaned up both the code
and the running binary size (the module is less than half the size of
the original).

Supported Cards/Chipsets (Limited release or deprecated cards preceded by a #)
-------------------------
	PCI ID (pci.ids)	OEM	Product
#	9005:0283:9005:0283	Adaptec	Catapult (3210S with arc firmware)
#	9005:0284:9005:0284	Adaptec	Tomcat (3410S with arc firmware)
	9005:0285:9005:0285	Adaptec	2200S (Vulcan)
	9005:0285:9005:0286	Adaptec	2120S (Crusader)
	9005:0285:9005:0287	Adaptec	2200S (Vulcan-2m)
	9005:0285:9005:0288	Adaptec	3230S (Harrier)
	9005:0285:9005:0289	Adaptec	3240S (Tornado)
	9005:0285:9005:028a	Adaptec	2020ZCR (Skyhawk)
	9005:0285:9005:028b	Adaptec	2025ZCR (Terminator)
	9005:0286:9005:028c	Adaptec	2230S (Lancer)
	9005:0286:9005:028c	Adaptec	2230SLP (Lancer)
	9005:0286:9005:028d	Adaptec	2130S (Lancer)
	9005:0285:9005:028e	Adaptec	2020SA (Skyhawk)
	9005:0285:9005:028f	Adaptec	2025SA (Terminator)
	9005:0285:9005:0290	Adaptec	2410SA (Jaguar)
	9005:0285:103c:3227	Adaptec	2610SA (Bearcat HP release)
#	9005:0285:0e11:0295	Adaptec	2610SA (Bearcat Compaq release)
	9005:0285:9005:0293	Adaptec	21610SA (Corsair-16)
	9005:0285:9005:0296	Adaptec	2240S (SabreExpress)
	9005:0285:9005:0292	Adaptec	2810SA (Corsair-8)
#	9005:0285:9005:0294	Adaptec	Prowler
	9005:0285:9005:0297	Adaptec	4005 (AvonPark)
	9005:0285:9005:0298	Adaptec	4000 (BlackBird)
	9005:0285:9005:0299	Adaptec	4800SAS (Marauder-X)
	9005:0285:9005:029a	Adaptec	4805SAS (Marauder-E)
	9005:0286:9005:029b	Adaptec	2820SA (Intruder)
	9005:0286:9005:029c	Adaptec	2620SA (Intruder)
	9005:0286:9005:029d	Adaptec	2420SA (Intruder HP release)
#	9005:0286:9005:02a2	Adaptec	3800 (Hurricane44)
#	9005:0286:9005:02a7	Adaptec	3805 (Hurricane80)
#	9005:0286:9005:02a8	Adaptec	3400 (Hurricane40)
	9005:0286:9005:02ac	Adaptec	1800 (Typhoon44)
#	9005:0286:9005:02b3	Adaptec	2400 (Hurricane40lm)
	9005:0285:9005:02b5	Adaptec	5445 (Voodoo44)
	9005:0285:15d9:02b5	SMC	AOC-USAS-S4i
	9005:0285:9005:02b6	Adaptec	5805 (Voodoo80)
	9005:0285:15d9:02b6	SMC	AOC-USAS-S8i
	9005:0285:9005:02b7	Adaptec	5085 (Voodoo08)
	9005:0285:9005:02bb	Adaptec	3405 (Marauder40LP)
	9005:0285:9005:02bc	Adaptec	3805 (Marauder80LP)
	9005:0285:9005:02c7	Adaptec	3085 (Marauder08ELP)
	9005:0285:9005:02bd	Adaptec	31205 (Marauder120)
	9005:0285:9005:02be	Adaptec	31605 (Marauder160)
	9005:0285:9005:02c3	Adaptec	51205 (Voodoo120)
	9005:0285:9005:02c4	Adaptec	51605 (Voodoo160)
	9005:0285:15d9:02c9	SMC	AOC-USAS-S4iR
	9005:0285:15d9:02ca	SMC	AOC-USAS-S8iR
	9005:0285:9005:02ce	Adaptec	51245 (Voodoo124)
	9005:0285:9005:02cf	Adaptec	51645 (Voodoo164)
	9005:0285:9005:02d0	Adaptec	52445 (Voodoo244)
	9005:0285:9005:02d1	Adaptec	5405 (Voodoo40)
	9005:0285:15d9:02d2	SMC	AOC-USAS-S8i-LP
	9005:0285:15d9:02d3	SMC	AOC-USAS-S8iR-LP
	9005:0285:9005:02d4	Adaptec	ASR-2045 (Voodoo04 Lite)
	9005:0285:9005:02d5	Adaptec	ASR-2405 (Voodoo40 Lite)
	9005:0285:9005:02d6	Adaptec	ASR-2445 (Voodoo44 Lite)
	9005:0285:9005:02d7	Adaptec	ASR-2805 (Voodoo80 Lite)
	9005:0285:9005:02d8	Adaptec	5405Z (Voodoo40 BLBU)
	9005:0285:9005:02d9	Adaptec	5445Z (Voodoo44 BLBU)
	9005:0285:9005:02da	Adaptec	5805Z (Voodoo80 BLBU)
#	9005:0285:9005:02db	Adaptec	5085G (Voodoo08 PM)
#	9005:0285:9005:02dc	Adaptec	51245G (Voodoo124 PM)
#	9005:0285:9005:02dd	Adaptec	51645G (Voodoo164 PM)
#	9005:0285:9005:02de	Adaptec	52445G (Voodoo244 PM)
#	9005:0285:9005:02df	Adaptec	ASR-2045G (Voodoo04 Lite PM)
#	9005:0285:9005:02e0	Adaptec	ASR-2405G (Voodoo40 Lite PM)
#	9005:0285:9005:02e1	Adaptec	ASR-2445G (Voodoo44 Lite PM)
#	9005:0285:9005:02e2	Adaptec	ASR-2805G (Voodoo80 Lite PM)
	1011:0046:9005:0364	Adaptec	5400S (Mustang)
	1011:0046:9005:0365	Adaptec	5400S (Mustang)
	9005:0287:9005:0800	Adaptec	Themisto (Jupiter)
	9005:0200:9005:0200	Adaptec	Themisto (Jupiter)
	9005:0286:9005:0800	Adaptec	Callisto (Jupiter)
	1011:0046:9005:1364	Dell	PERC 2/QC (Quad Channel, Mustang)
	1011:0046:9005:1365	Dell	PERC 2/QC (Quad Channel, Mustang)
	1028:0001:1028:0001	Dell	PERC 2/Si (Iguana)
	1028:0003:1028:0003	Dell	PERC 3/Si (SlimFast)
	1028:0002:1028:0002	Dell	PERC 3/Di (Opal)
	1028:0004:1028:0004	Dell	PERC 3/SiF (Iguana)
	1028:0004:1028:00d0	Dell	PERC 3/DiF (Iguana)
	1028:0002:1028:00d1	Dell	PERC 3/DiV (Viper)
	1028:0002:1028:00d9	Dell	PERC 3/DiL (Lexus)
	1028:000a:1028:0106	Dell	PERC 3/DiJ (Jaguar)
	1028:000a:1028:011b	Dell	PERC 3/DiD (Dagger)
	1028:000a:1028:0121	Dell	PERC 3/DiB (Boxster)
	9005:0285:1028:0287	Dell	PERC 320/DC (Vulcan)
	9005:0285:1028:0291	Dell	CERC 2 (DellCorsair)
	1011:0046:103c:10c2	HP	NetRAID-4M (Mustang)
	9005:0285:17aa:0286	Legend	S220 (Crusader)
	9005:0285:17aa:0287	Legend	S230 (Vulcan)
	9005:0285:9005:0290	IBM	ServeRAID 7t (Jaguar)
	9005:0285:1014:02F2	IBM	ServeRAID 8i (AvonPark)
#	9005:0285:1014:0312	IBM	ServeRAID 8i (AvonParkLite)
	9005:0286:1014:9540	IBM	ServeRAID 8k/8k-l4 (AuroraLite)
	9005:0286:1014:9580	IBM	ServeRAID 8k/8k-l8 (Aurora)
#	9005:0286:1014:034d	IBM	ServeRAID 8s (Hurricane)
	9005:0285:1014:034d	IBM	ServeRAID 8s (Marauder-E)
	9005:0286:9005:029e	ICP	ICP9024RO (Lancer)
	9005:0286:9005:029f	ICP	ICP9014RO (Lancer)
	9005:0286:9005:02a0	ICP	ICP9047MA (Lancer)
	9005:0286:9005:02a1	ICP	ICP9087MA (Lancer)
#	9005:0286:9005:02a3	ICP	ICP5445AU (Hurricane44)
	9005:0285:9005:02a4	ICP	ICP9085LI (Marauder-X)
	9005:0285:9005:02a5	ICP	ICP5085BR (Marauder-E)
	9005:0286:9005:02a6	ICP	ICP9067MA (Intruder-6)
#	9005:0286:9005:02a9	ICP	ICP5085AU (Hurricane80)
#	9005:0286:9005:02aa	ICP	ICP5045AU (Hurricane40)
	9005:0285:9005:02b2	ICP	(Voodoo 8 internal 8 external)
#	9005:0286:9005:02b4	ICP	ICP5045AL (Hurricane40lm)
	9005:0285:9005:02b8	ICP	ICP5445SL (Voodoo44)
	9005:0285:9005:02b9	ICP	ICP5085SL (Voodoo80)
	9005:0285:9005:02ba	ICP	ICP5805SL (Voodoo08)
	9005:0285:9005:02bf	ICP	ICP5045BL (Marauder40LP)
	9005:0285:9005:02c0	ICP	ICP5085BL (Marauder80LP)
	9005:0285:9005:02c8	ICP	ICP5805BL (Marauder08ELP)
	9005:0285:9005:02c1	ICP	ICP5125BR (Marauder120)
	9005:0285:9005:02c2	ICP	ICP5165BR (Marauder160)
	9005:0285:9005:02c5	ICP	ICP5125SL (Voodoo120)
	9005:0285:9005:02c6	ICP	ICP5165SL (Voodoo160)
	9005:0286:9005:02ab		(Typhoon40)
	9005:0286:9005:02ad		(Aurora ARK)
	9005:0286:9005:02ae		(Aurora Lite ARK)
	9005:0285:9005:02b0		(Sunrise Lake ARK)
	9005:0285:9005:02b1	Adaptec	(Voodoo 8 internal 8 external)
	9005:0285:108e:7aac	SUN	STK RAID REM (Voodoo44 Coyote)
	9005:0285:108e:0286	SUN	STK RAID INT (Cougar)
	9005:0285:108e:0287	SUN	STK RAID EXT (Prometheus)
	9005:0285:108e:7aae	SUN	STK RAID EM (Narvi)

People
-------------------------
Alan Cox <alan@redhat.com>
Christoph Hellwig <hch@infradead.org>	(updates for new-style PCI probing and SCSI host registration,
					 small cleanups/fixes)
Matt Domsch <matt_domsch@dell.com>	(revision ioctl, adapter messages)
Deanna Bonds                            (non-DASD support, PAE fibs and 64 bit, added new adaptec controllers
					 added new ioctls, changed scsi interface to use new error handler,
					 increased the number of fibs and outstanding commands to a container)

					(fixed 64bit and 64G memory model, changed confusing naming convention
					 where fibs that go to the hardware are consistently called hw_fibs and
					 not just fibs like the name of the driver tracking structure)
Mark Salyzyn <Mark_Salyzyn@adaptec.com> Fixed panic issues and added some new product ids for upcoming hbas. Performance tuning, card failover and bug mitigations.
Achim Leubner <Achim_Leubner@adaptec.com>

Original Driver
-------------------------
Adaptec Unix OEM Product Group

Mailing List
-------------------------
linux-scsi@vger.kernel.org (Interested parties troll here)
Also note this is very different to Brian's original driver
so don't expect him to support it.
Adaptec does support this driver.  Contact Adaptec tech support or
aacraid@adaptec.com

Original by Brian Boerner February 2001
Rewritten by Alan Cox, November 2001


CHANGELOG
Aacraid (ARC) Linux driver release notes. Ascending (chronological) order
comments. Modifications performed by Mark_Salyzyn@adaptec.com and 
Achim_Leubner@adaptec.com if not otherwise noted. 
Document created 2003-05-15 at Version 0.9.10.

Streams support otc-main, avonpark and enzo

Version: 0.9.10

Version: 1.1.2

2003-05-15	Mark_Salyzyn@adaptec.com

Differences between 2.4.21-rc2-ac2 kernel and our 1.1.2 versioned driver,
changes as performed by Deanna Bonds, Bob Pasteur and Mark Salyzyn.

aachba.c:
	- If the state of a logical unit is hidden, then do not report. This
	  state is typically entered when a device is being cleared.
	- Added support for the Tallahassee project, where one channel is
	  dedicated to SCSI, and the other channel is dedicated to RAID.
	- Resolved some issues surrounding PAE support and IA64.
	- If the driver is a not a boot disk driver, then set the Removable
	  bit on the inquiry strings returned by the logical units to ensure
	  that any changes in the arrays will be acquired when the device is
	  re-attached.
	- mask the SRB status with 0x3F to deal with misbehaving devices.
	- Do not report DISKs to inquiry requests on the SCSI bus except if
	  the channel is designated as a SCSI only bus.
	- Propagate check conditions to the SCSI command result.
	- Add support for programmable timeouts to propagate down toe the
	  requests.
	- If we have pae mode enabled, right after we get the adapter
	  information and determine the pae mode capability, we enable the
	  system to issue 64 bit requests.
aacraid.h:
	- Had to drop from 512 commands to 100 commands because some versions
	  of the firmware would starve commands causing a timeout reaction
	  which lead to lost commands.
	- Added a global control variable for nondasd and paemode support.
	- Dealt with some 64 bit / 32 bit issues in list_head structures and
	  helper Macros, replacing them with our own more sensitive variants.
	- Differentiated virtual and physical references to the shared fib
	  allocations.
	- information structure not synchronized to firmware, needed to add
	  a clusterchannelmask.
	- Added definitions in support of the new configuration information
	  page bits in support of Tallahassee.
	- Changed to an allocated fib pool, rather than an array in the hba
	  structure as this affected the SCSI memory pool.
	- Added some AIF definitions to permit us to sniff for container
	  changes to permit a rescan to pick up new information or targets.
commctrl.c:
	- The fib reference was changed to a physical and a virtual address,
	  absorb the name changes.
	- The list_head structure handlers have been replaced with our own,
	  absorb the name changes.
	- The fib address reported in an AIF is a physical (32 bit) reference,
	  and not a virtual (possibly 64 bit) reference.
	- added the ioctl handling for sending a raw srb (FSACTL_SEND_RAW_SRB).
comminit.c:
	- Deal with IA64 issues.
	- Change to using the physical address (32 bit) for the AIF references.
	- The list_head structure handlers have been replaced with our own,
	  absorb the name changes.
	- Observed a memory leak, free up the queue resources should we fail
	  to initialize the adapter.
commsup.c:
	- The fib reference was changed to a physical and a virtual address,
	  absorb the name changes.
	- Instead of panicking the kernel when a fib allocation was available,
	  sleep until it is available.
	- Submitted fib pointers are physical (32 bit) rather than virtual
	  (possibly 64 bit) values.
	- producer and consumer indexes should be converted over to local
	  cpu endian before comparison.
	- aac_handle_aif now sniffs AIF events and takes plug and play action
	  for container changes.
	- The aif thread is set up to be a kernel thread, and not a user
	  thread. This permits us the ability to make plug and play calls
	  without prejudice.
	- Added instrumentation to the aif thread to confirm the plug and
	  play activity and as an aid to several other debug sessions.
	- Do not age an aif context based on the last received aif, but rather
	  the last poll.
dpcsup.c:
	- The fib reference was changed to a physical and a virtual address,
	  absorb the name changes.
	- Submitted fib pointers are physical (32 bit) rather than virtual
	  (possibly 64 bit) values.
linit.c:
	- Added paemode control.
	- Added various upcoming board products, and documented better the
	  existing board product ids. This includes SATA RAID products.
	- needed to take the io_request_lock during portions of initialization.
	- allocate the fib resource separately, rather than part of adapter
	  structure to aid in the precious SCSI resources.
	- cleanup of none dasd support options.
	- Added more details about the build date of the driver to the proc
	  information.
	- dropped a change that permitted 64 bit DMA resources to be generated
	  instead of through a 32 bit bounce buffer. (it was moved to aachba.c
	  where it can be turned on after we determine the adapter's
	  capabilities).
	- max_id, max_lun and max_channel parameters are set after the
	  adapter information has been picked up (the number of channels is
	  based on the product id table now).
sa.c:
	- Context of timeout handling was incorrect, only noticed in IA64
	  bit machines (due to lack of BIOS initialization).

Differences that need further investigation and could be viewed as regressions
and added after submission:

rx.c:
	- Dropped detection of failure to generate kernel command thread.
sa.c:
	- Dropped detection of failure to generate kernel command thread.

Version: 1.1.3

2003-07-01	Mark_Salyzyn@adaptec.com

aachba.c:
	- Added aac_get_container_name to permit override of array inquiry
	  string with the set name.

2003-07-08	Mark_Salyzyn@adaptec.com

aachba.c:
	- Return 0 (success) for unsupported commands, the check condition
	  should perform the necessary action of error handling.

2003-07-10	Mark_Salyzyn@adaptec.com

aachba.c:
	- The pass-through SCSI SCB command in PAE mode was getting the fib

compat.h
/*
 *	Adaptec AAC series RAID controller driver
 *	(c) Copyright 2001 Red Hat Inc.	<alan@redhat.com>
 *
 * Copyright (c) 2000-2010 Adaptec, Inc. (aacraid@adaptec.com)
 * Copyright (c) 2010-2015 PMC-Sierra, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 * 
 * Module Name:
 *   compat.h
 *
 *
 * Abstract: This file is for backwards compatibility with older kernel versions
 */
#include <linux/version.h>

#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,15))
#ifndef sdev_channel
#define sdev_channel(x)	(x)->channel
#endif
#ifndef scmd_channel
#define scmd_channel(x)	sdev_channel((x)->device)
#endif
#ifndef sdev_id
#define sdev_id(x)	(x)->id
#endif
#ifndef scmd_id
#define scmd_id(x)	sdev_id((x)->device)
#endif

#endif
#if (LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,22))
#ifndef scsi_sglist
#define scsi_sglist(cmd) ((struct scatterlist *)((cmd)->request_buffer))
#endif
#ifndef scsi_bufflen
#define scsi_bufflen(cmd) ((cmd)->request_bufflen)
#endif
#ifndef scsi_sg_count
#define scsi_sg_count(cmd) ((cmd)->use_sg)
#endif
#ifndef SCSI_HAS_DMA_MAP
#define scsi_dma_unmap(cmd) if(scsi_sg_count(cmd))pci_unmap_sg(((struct aac_dev *)cmd->device->host->hostdata)->pdev,scsi_sglist(cmd),scsi_sg_count(cmd),cmd->sc_data_direction)
#if ((defined(__VMKERNEL_MODULE__) || defined(__VMKLNX30__)) && !defined(__VMKLNX__))
#define scsi_dma_map(cmd) scsi_sg_count(cmd)
#else
#define scsi_dma_map(cmd) ((scsi_sg_count(cmd))?pci_map_sg(((struct aac_dev *)cmd->device->host->hostdata)->pdev,scsi_sglist(cmd),scsi_sg_count(cmd),cmd->sc_data_direction):0)
#endif
#endif
#if (!defined(__VMKLNX__))
#define sg_next(sg) ((sg)+1)
#endif
#ifndef scsi_resid
#define scsi_resid(cmd) ((cmd)->resid)
#define scsi_set_resid(cmd,res) (cmd)->resid = res
#endif
#ifndef scsi_for_each_sg
#define scsi_for_each_sg(cmd, sg, nseg, __i) \
	for (__i = 0, sg = scsi_sglist(cmd); __i < (nseg); __i++, sg = sg_next(sg))
#endif

#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,24))
# define uintptr_t ptrdiff_t
#endif
#if ((LINUX_VERSION_CODE < KERNEL_VERSION(2,6,9)) && !defined(SCSI_HAS_SSLEEP))
#undef ssleep
#define ssleep(x) scsi_sleep((x)*HZ)
#endif
#if ((LINUX_VERSION_CODE < KERNEL_VERSION(2,6,7)) && !defined(HAS_MSLEEP))
#define msleep(x) set_current_state(TASK_UNINTERRUPTIBLE); schedule_timeout(x)
#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,4,19))
#ifndef __builtin_expect
#define __builtin_expect(x,expected_value) (x)
#endif
#ifndef unlikely
#define unlikely(x) __builtin_expect(!!(x),0)
#endif
#ifndef likely
#define likely(x) __builtin_expect(!!(x),1)
#endif
#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,9))
#ifndef BUG_ON
#define BUG_ON(condition) do { if (unlikely((condition)!=0)) BUG(); } while (0)
#endif
#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,4,10))
#ifndef min
#define min(a,b) (((a)<(b))?(a):(b))
#endif
#ifndef min_t
#define min_t(a,b,c) (((a)(b)<(a)(c))?(a)(b):(a)(c))
#endif
#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,4,0))
 typedef unsigned long dma_addr_t;
#include <linux/kcomp.h>
#define PCI_ANY_ID (~0)
#define SCSI_DATA_UNKNOWN	0
#define SCSI_DATA_WRITE		1
#define SCSI_DATA_READ		2
#define SCSI_DATA_NONE		3
 /* Sigh ... a *lot* more needs to be done for this Grandpa */

#endif

#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,5,0))
#include <linux/time.h>
static inline unsigned long get_seconds(void)
{
	struct timeval now;
	do_gettimeofday(&now);
	return now.tv_sec;
}
#define scsi_host_template SHT
#define DMA_BIDIRECTIONAL	SCSI_DATA_UNKNOWN
#define DMA_TO_DEVICE		SCSI_DATA_WRITE
#define DMA_FROM_DEVICE		SCSI_DATA_READ
#define DMA_NONE		SCSI_DATA_NONE
#if (defined(__ESXi4__))
#define imajor(x) MAJOR(x->i_rdev) //ESXi4 support
#else
#define iminor(x) MINOR(x->i_rdev)
#endif
#define scsi_host_alloc(t,s) scsi_register(t,s)
#if (defined(CONFIG_VMNIX) && (LINUX_VERSION_CODE < KERNEL_VERSION(2,4,4)))
#define scsi_host_put(s)
#else
#define scsi_host_put(s) scsi_unregister(s)
#endif
#ifndef pci_set_consistent_dma_mask
#define pci_set_consistent_dma_mask(d,m) 0
#endif
#ifndef SCSI_HAS_SCSI_SCAN_HOST
#ifndef scsi_scan_host
#define scsi_scan_host(s)
#endif
#endif
#define scsi_add_host(s,d) 0
#if ((LINUX_VERSION_CODE < KERNEL_VERSION(2,4,20)) && !defined(list_for_each_entry))
#if (!defined(_LINUX_PREFETCH_H))
static inline void prefetch(const void *x) {;}
#endif
#define list_for_each_entry(pos, head, member)                          \
        for (pos = list_entry((head)->next, typeof(*pos), member),      \
                     prefetch(pos->member.next);                        \
             &pos->member != (head);                                    \
             pos = list_entry(pos->member.next, typeof(*pos), member),  \
                     prefetch(pos->member.next))
#endif
#if (defined(__VMKERNEL_MODULE__) || defined(__VMKLNX30__) || defined(__VMKLNX__))
#define scsi_remove_host(s)
#else
#define scsi_remove_host(s)					\
	struct proc_dir_entry * entry = NULL;			\
	struct scsi_device *device;				\
	extern struct proc_dir_entry * proc_scsi;		\
	if (proc_scsi != (struct proc_dir_entry *)NULL)		\
	for (entry = proc_scsi->subdir;				\
	  entry != (struct proc_dir_entry *)NULL &&		\
	  (!entry->low_ino ||					\
	    (entry->namelen != 4) ||				\
	    memcmp ("scsi", entry->name, 4));			\
	  entry = entry->next);					\
	if (entry && entry->write_proc)				\
	for (device = s->host_queue;				\
	  device != (struct scsi_device *)NULL;			\
	  device = device->next)				\
		if (!device->access_count && !s->in_recovery) {	\
			char buffer[80];			\
			int length;				\
			mm_segment_t fs;			\
			device->removable = 1;			\
			sprintf (buffer, "scsi "		\
			  "remove-single-device %d %d %d %d\n", \
			  s->host_no, device->channel,		\
			  device->id, device->lun);		\
			length = strlen (buffer);		\
			fs = get_fs();				\
			set_fs(get_ds());			\
			length = entry->write_proc(		\
			  NULL, buffer, length, NULL);		\
			set_fs(fs);				\
		}
#endif
#if (!defined(__devexit_p))
# if (defined(MODULE))
#  define __devexit_p(x) x
# else
#  define __devexit_p(x) NULL
# endif
#endif
#define __user
#endif
#if ((LINUX_VERSION_CODE < KERNEL_VERSION(2,6,6)) && (!defined(SCSI_HAS_SCSI_DEVICE_ONLINE)))
#define scsi_device_online(d) ((d)->online)
#ifndef SDEV_RUNNING
#define SDEV_RUNNING 1
#endif
#ifndef SDEV_OFFLINE
#define SDEV_OFFLINE 0
#endif
#define scsi_device_set_state(d,s) ((d)->online=(s!=SDEV_OFFLINE))
#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,9))
#define __iomem
#endif
#if ((LINUX_VERSION_CODE < KERNEL_VERSION(2,6,9)) && !defined(HAS_BITWISE_TYPE))
typedef u64 __le64;
typedef u32 __le32;
typedef u16 __le16;

#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,13))
#if (LINUX_VERSION_CODE > KERNEL_VERSION(2,5,0))
#include <linux/dma-mapping.h>
#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,5))
#ifndef DMA_64BIT_MASK
#define DMA_64BIT_MASK ((dma_addr_t)0xffffffffffffffffULL)
#endif
#ifndef DMA_32BIT_MASK
#define DMA_32BIT_MASK ((dma_addr_t)0xffffffffULL)
#endif
#endif
#ifndef DMA_31BIT_MASK
#define DMA_31BIT_MASK ((dma_addr_t)0x7fffffffULL)
#endif
#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(3,0,0))
#define DMA_BIT_MASK(n) (((n) == 64) ? ~0ULL : ((1ULL<<(n))-1))
#endif
#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,8,0))
#define __devinit
#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,11))
#ifndef spin_trylock_irqsave
#define spin_trylock_irqsave(lock, flags) \
({ \
	local_irq_save(flags); \
	spin_trylock(lock) ? \
	1 : ({local_irq_restore(flags); 0 ;}); \
})
#endif

#endif
#if ((LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,0)) && (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,1)) && defined(SCSI_HAS_MY_DEVICES))
#define shost_for_each_device(sdev, shost) \
	for ((sdev) = __scsi_iterate_devices((shost), NULL); \
	     (sdev); \
	     (sdev) = __scsi_iterate_devices((shost), (sdev)))
#define __shost_for_each_device(sdev, shost) \
	list_for_each_entry((sdev), &((shost)->my_devices), siblings)
#include <scsi/scsi_host.h>
#include <scsi/scsi_device.h>
static inline struct scsi_device *__scsi_iterate_devices(struct Scsi_Host *shost,
			                   struct scsi_device *prev)
{
	struct list_head *list = (prev ? &prev->siblings : &shost->my_devices);
	struct scsi_device *next = NULL;
	unsigned long flags;

	spin_lock_irqsave(shost->host_lock, flags);
	while (list->next != &shost->my_devices) {
		next = list_entry(list->next, struct scsi_device, siblings);
		/* skip devices that we can't get a reference to */
		if (!scsi_device_get(next))
			break;
		list = list->next;
	}
	spin_unlock_irqrestore(shost->host_lock, flags);

	if (prev)
		scsi_device_put(prev);
	return next;
}

#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,4,18)) && (LINUX_VERSION_CODE != KERNEL_VERSION(2,4,9)) && (LINUX_VERSION_CODE != KERNEL_VERSION(2,4,13))
# define dma_handle ptr

#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,4,11))
#include <linux/blk.h>

static inline unsigned int block_size(kdev_t dev)
{
	int retval = BLOCK_SIZE;
	int major = MAJOR(dev);

	if (blksize_size[major]) {
		int minor = MINOR(dev);
		if (blksize_size[major][minor])
			retval = blksize_size[major][minor];
	}
	return retval;
}

#endif

/*
 * ESXI4 does not support list_is_singular
 */

#if (defined(RHEL_MAJOR) && RHEL_MAJOR == 5 && RHEL_MINOR <= 2)
#define RHEL5_2
#endif

#if (defined(__ONLY_ESXi4__)) || defined(RHEL5_2) || (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,18) && !(defined(__VMKLNX__)))
 static inline int list_is_singular(const struct list_head *head)
{
         return !list_empty(head) && (head->next == head->prev);
}
#endif

/*
 * Vmware does not suppot get_unaligned_be64
 */

#if (defined(__VMKLNX__)) || (LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,18))
static inline u32 __get_unaligned_be32(const u8 *p)
{
	return p[0] << 24 | p[1] << 16 | p[2] << 8 | p[3];

}

static inline u32 get_unaligned_be32(const void *p)
{
        return __get_unaligned_be32((const u8 *)p);
}

static inline u64 __get_unaligned_be64(const u8 *p)
{
         return (u64)__get_unaligned_be32(p) << 32 |
                __get_unaligned_be32(p + 4);
}

static inline u64 get_unaligned_be64(const void *p)
{
         return __get_unaligned_be64((const u8 *)p);
}

#endif

#if (LINUX_VERSION_CODE <= KERNEL_VERSION(2,4,4))
# define pci_disable_device(x)
#endif
#if (LINUX_VERSION_CODE <= KERNEL_VERSION(2,4,7))

#ifndef COMPLETION_INITIALIZER

#include <linux/wait.h>

struct completion {
	unsigned int done;
	wait_queue_head_t wait;
};
#define COMPLETION_INITIALIZER(work) \
	{ 0, __WAIT_QUEUE_HEAD_INITIALIZER((work).wait) }

#define DECLARE_COMPLETION(work) \
	struct completion work = COMPLETION_INITIALIZER(work)
#define INIT_COMPLETION(x)	((x).done = 0)

static inline void init_completion(struct completion *x)
{
	x->done = 0;
	init_waitqueue_head(&x->wait);
}
#endif

#ifndef complete_and_exit
static inline void complete_and_exit(struct completion *comp, long code)
{
	/*
	if (comp)
		complete(comp);

	do_exit(code);
	*/
}
#endif

#endif
#if ((LINUX_VERSION_CODE < KERNEL_VERSION(2,6,23)) && !defined(sdev_printk))
#define sdev_printk(prefix, sdev, fmt, a...) \
	printk(prefix " %d:%d:%d:%d: " fmt, sdev->host->host_no, \
		sdev_channel(sdev), sdev_id(sdev), sdev->lun, ##a)
#endif
#if (LINUX_VERSION_CODE <= KERNEL_VERSION(2,4,2))
#include <linux/pci.h>

//static inline void scsi_set_pci_device(struct Scsi_Host *SHpnt,
//                                       struct pci_dev *pdev)
//{
///*	SHpnt->pci_dev = pdev;	*/
//}
#define scsi_set_pci_device(SHpnt,pdev)

static inline void wait_for_completion(struct completion *x)
{
	spin_lock_irq(&x->wait.lock);
	if (!x->done) {
		DECLARE_WAITQUEUE(wait, current);

#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,4,0))
		wait.flags |= WQ_FLAG_EXCLUSIVE;
#endif
		__add_wait_queue_tail(&x->wait, &wait);
		do {
			__set_current_state(TASK_UNINTERRUPTIBLE);
			spin_unlock_irq(&x->wait.lock);
			schedule();
			spin_lock_irq(&x->wait.lock);
		} while (!x->done);
		__remove_wait_queue(&x->wait, &wait);
	}
	x->done--;
	spin_unlock_irq(&x->wait.lock);
}

static inline int pci_set_dma_mask(struct pci_dev *dev, dma_addr_t mask)
{
#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,4,0))
    dev->dma_mask = mask;
#endif

    return 0;
}

#endif
#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,18))

#ifndef IRQF_SHARED
# define IRQF_SHARED SA_SHIRQ
#endif
#ifndef IRQF_DISABLED
# define IRQF_DISABLED SA_INTERRUPT /* Counter intuitive? */
#endif
#endif
